test/Tests/src/Data/Range_Spec.enso

```diff
@@ -56,4 +80,217 @@ spec = Test.group "Range" <|
+        r5.find (x-> x*x == 25) . should_equal 5
+        r5.find (x-> x*x == 4) . should_equal Nothing
+        verify_contains r5 [5] [0, 1, 4, 6, 7, 10]
+
+    Test.specify "should behave correctly with negative step" <|
+        r1 = Range 4 0 -1
+        r1.is_empty . should_be_false
+        r1.not_empty . should_be_true
+        r1.to_vector . should_equal [4, 3, 2, 1]
+        build_with_each r1 . should_equal [4, 3, 2, 1]
+        r1.map x->x+1 . should_equal [5, 4, 3, 2]
+        r1.fold 0 (+) .  should_equal 10
+        r1.any (_ == 2) . should_equal True
+        r1.any (_ == 0) . should_equal False
+        r1.all (x-> x % 2 == 0) . should_equal False
+        r1.all (_ > 0) . should_equal True
+        r1.find (x-> x*x == 16) . should_equal 4
+        r1.find (x-> x*x == 0) . should_equal Nothing
+        verify_contains r1 [4, 3, 2, 1] [-2, -1, 0, 5, 6, 7, 10]
+
+        r2 = Range 4 0 -2
+        r2.is_empty . should_be_false
+        r2.not_empty . should_be_true
+        r2.to_vector . should_equal [4, 2]
+        build_with_each r2 . should_equal [4, 2]
+        r2.map x->x+1 . should_equal [5, 3]
+        r2.fold 0 (+) .  should_equal 6
+        r2.any (_ == 2) . should_equal True
+        r2.any (_ == 0) . should_equal False
+        r2.all (x-> x % 2 == 0) . should_equal True
+        r2.all (_ > 2) . should_equal False
+        r2.find (x-> x*x == 16) . should_equal 4
+        r2.find (x-> x*x == 0) . should_equal Nothing
+        verify_contains r2 [4, 2] [-2, -1, 0, 1, 3, 5, 6, 7, 10]
+
+        r3 = Range 4 0 -10
+        r3.is_empty . should_be_false
+        r3.not_empty . should_be_true
+        r3.to_vector . should_equal [4]
+        build_with_each r3 . should_equal [4]
+        r3.map x->x+1 . should_equal [5]
+        r3.fold 0 (+) .  should_equal 4
+        r3.any (_ == 2) . should_equal False
+        r3.any (_ == 4) . should_equal True
+        r3.all (x-> x % 2 == 0) . should_equal True
+        r3.all (_ > 4) . should_equal False
+        r3.find (x-> x*x == 16) . should_equal 4
+        r3.find (x-> x*x == 0) . should_equal Nothing
+        verify_contains r3 [4] [-2, -1, 0, 1, 2, 3, 5, 6, 7, 10]
+
+        r4 = Range 3 0 -3
+        r4.is_empty . should_be_false
+        r4.not_empty . should_be_true
+        r4.to_vector . should_equal [3]
+        build_with_each r4 . should_equal [3]
+        r4.map x->x+1 . should_equal [4]
+        r4.fold 0 (+) .  should_equal 3
+        r4.any (_ == 3) . should_equal True
+        r4.any (_ == 2) . should_equal False
+        r4.all (x-> x % 2 == 0) . should_equal False
+        r4.all (_ > 0) . should_equal True
+        r4.find (x-> x*x == 9) . should_equal 3
+        r4.find (x-> x*x == 0) . should_equal Nothing
+        verify_contains r4 [3] [-3, -2, -1, 0, 1, 2, 4, 5, 6, 7, 10]
+
+    Test.specify "should report errors if trying to set step to 0" <|
+        0.up_to 0 . with_step 0 . should_fail_with Illegal_State_Error
+        invalid_range = Range 0 0 0
```

Comment:
Maybe test `Range 0 0 2` as well? Just to cover corner cases

Reply:
But this range is not invalid, it is just empty.

Reply:
A range is only invalid if its step is 0 - because the step should be a non-zero integer (so technically a range is also invalid if the step is decimal, but this is a typelevel check).

A range whose step exceeds `end-start` is still valid, but will be empty if `start=end` or contain just a single element `[start]`.